feat: allow pod IPs even for non-hostNetwork pods

peterhoneder opened the pull request:
but only for the internalHostnameAnnotationKey which doesn't use the node's IP anyway

<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

Pod IPs and node IPs can be used for hostname annotations on pods.

The existing implementation worked this way:
- check if the pod has nodeNetwork: true
- if no -> skip
- otherwise:
  - for the internal hostname annotation -> use the pod' IP
  - for the regular hostname annotation -> use the node's IP
  - special handling for kops-dns-controller

I would like to change the implementation to cover a case, where Pod IPs can be exposed as names, but with pods where hostNetwork == false. This is a typical use case with CNI plugins like Calico CNI, where pod IPs are exposed as BGP routes directly to the ToR router/switch. So there is basically no service in front of the pod and the pod is directly exposed to the outside world of the cluster, most commonly with an internet-routable public IP, but could also be a private IP that is justed routed internally. The usefulness of this change addresses both those cases.

Since the pod's IP is already used for the internal hostname annotation, there is no need to actually limit this behavor to pods with nodeNetwork: true.

I can add public documentation for this, since the whole pod annotation part is not publicly documented yet, it might be a topic for a separate PR though.

There is no github issue for this yet.

**Checklist**

- [x] Unit tests updated
- [ ] End user documentation updated

johngmyers commented:
It looks like dns-controller is incorrectly emulating kOps dns-controller. The `dns.alpha.kubernetes.io/internal` annotation on a pod causes dns-controller to register the host-network pod with each of its node's internal IP addresses. This is different than registering with the pod IP.

Both cases are useful. In some networking configurations, the pods' internal IPs are inaccessible, yet the node internal IPs are.

peterhoneder commented:
> kOps dns-controller

the proposed PR currently does not change that behavior, but yes the existing implementation of the non kops compatible code path would only use NodeIPs from host network pods

johngmyers commented:
I have a vague recollection of pods created in the early minutes of a new cluster having odd IP addresses, but I can't find anything like that in e2e tests. In the tests I've looked over, `podIPs` of host-network pods are the internal IPs of the node. So there might not be a distinction after all.

Whether or not non-host-network pods are included can be controlled by whether non-host-network pods have the annotation.

There's a good question of whether it should use `podIP` or `podIPs`, but that's for after external-dns supports AAAA records.

/lgtm

johngmyers commented:
I think I remember what it was: before the CNI comes up, the `podIP` field might not get populated. So if your CNI depends on external-dns registering the kube-apiserver pods with DNS, you need it to grab the IPs from the pod's node.

As previously noted, this was broken before and remains broken after the PR. The fix would be to follow from the pod to the node for pods that are host-networking. Fixing external-dns to correctly implement kOps requirements is behind #2051.

peterhoneder commented:
Hi!

my change does not affect the behavior for host-networking. It only allows an additional behavior for *non* host-networking pods, where the PodIP *is* correctly populated, and must be used instead of the host IP. The typical use case would be with a CNI like calico, where e.g. publicly routable IPv4 addresses are assigned to the pod.

johngmyers commented:
To be clear, I was not asking this PR to fix the handling of kOps requirements. It does not make things worse.

peterhoneder commented:
> To be clear, I was not asking this PR to fix the handling of kOps requirements. It does not make things worse.

ok, thanks, perfect 😄 just wanted to make sure we have the same understanding 👍

szuecs commented on source/pod_test.go:
```diff
@@ -254,7 +254,7 @@ func TestPodSource(t *testing.T) {
 			[]*endpoint.Endpoint{
 				{DNSName: "a.foo.example.org", Targets: endpoint.Targets{"54.10.11.1"}, RecordType: endpoint.RecordTypeA},
-				{DNSName: "internal.a.foo.example.org", Targets: endpoint.Targets{"10.0.1.1"}, RecordType: endpoint.RecordTypeA},
+				{DNSName: "internal.a.foo.example.org", Targets: endpoint.Targets{"10.0.1.1", "100.0.1.2"}, RecordType: endpoint.RecordTypeA},
```
The test name seems to be wrong.

peterhoneder commented:
> @peterhoneder Do you think you can rebase this PR and takes into account review comment from @szuecs ?

yes, I will do that

peterhoneder commented:
> @peterhoneder Do you think you can rebase this PR and takes into account review comment from @szuecs ?

I rebased the PR and added more tests and more explicit comments on why the tests are there that way. I tested the scenarios described in the commit message.

Regarding the comments further up in the thread especially from @johngmyers I think this points to valid use cases that might be interesting. In our setup we are basically using the internal hostname annotation for pods where calico CNI provides addresses and later announces those over BGP. Those addresses are public addresses due to the nature of the workload. In those cases having the regular hostname annotation work similarly in cases where the pod is not in the host network would be beneficial, but due to the constraints described above this sounds to me more like a task for a future PR.

mloiseleur commented:
/ok-to-test

mloiseleur commented:
/lgtm

mloiseleur commented on source/pod.go:
```diff
@@ -104,43 +99,46 @@ func (ps *podSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, error
-						// IPv6 addresses are labeled as NodeInternalIP despite being usable externally as well.
-						if address.Type == corev1.NodeExternalIP || (address.Type == corev1.NodeInternalIP && recordType == endpoint.RecordTypeAAAA) {
-							addToEndpointMap(endpointMap, domain, recordType, address.Address)
+		// accept annotations that point to the node's external IP (or IPv6 NodeInternalIP) only for pods using the host network
+		if pod.Spec.HostNetwork {
```
Wdyt about removing this condition to cover all cases described in https://github.com/kubernetes-sigs/external-dns/pull/3468#issuecomment-1537556184 ?

peterhoneder replied:
I can include that change in here, will do so later today.

peterhoneder replied:
@mloiseleur please check, I added the changes and also some more docs updates.

mloiseleur replied:
It's lgtm at first glance. I asked @gregorycuellar to also check, if possible.

mloiseleur commented:
@peterhoneder improved this PR to include this use case. Wdyt ? Anything missing ?

gregorycuellar commented:
@mloiseleur It look good to me, I think it's ok

mloiseleur commented:
/lgtm
/assign @szuecs

omartheironman commented:
is there an update on this?

peterhoneder commented:
Hi! any updates on this?

perber commented:
Hi! any updates on this?

szuecs commented:
I don't think DNS is a good choice for this use case. They should rather create a service type loadbalancer. Maybe calico can build a svc type loadbalancer if they see value in supporting it.

In general I would rather delete pod dns name functionality than invest more into it.

peterhoneder commented:
> I don't think DNS is a good choice for this use case. They should rather create a service type loadbalancer. Maybe calico can build a svc type loadbalancer if they see value in supporting it.
> 
> In general I would rather delete pod dns name functionality than invest more into it.

Let me explain the use case: there are services which cannot be put behind a service. E.g. if you have multiple connections which need to end up on the same replica, and are based on DNS round robin, a k8s service would not help. It would only help if a service would be created for each pod separately. The patch works very well in production and seems to be used by multiple people who even created multiple PRs for this.

jcralbino commented:
Hello All,

I am a user of external DNS, and I manage, design and operate enterprise container platform.

I will like to clarify some assumptions from this thread, and state that any POD IP can be reached for external communication if they are routed to the network.

https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid-Integrated-Edition/1.19/tkgi/GUID-network-profiles-pods.html


That is what we use for our enterprise deployment, routed pod IPs, where SNAT is not used and all communication between/to a pod can be allowed if the security policy allows it.

It is my understanding that using this check for hostNetwork=False, and assuming that those pods should be ignored is also wrong

jcralbino commented:
> I don't think DNS is a good choice for this use case. They should rather create a service type loadbalancer. Maybe calico can build a svc type loadbalancer if they see value in supporting it.
> 
> In general I would rather delete pod dns name functionality than invest more into it.

I believe that use case is indeed not the only one.

Pod IPs can be routable, and allowing it to have a dns name can be relevant for many different scenarios

didlawowo commented:
any news ?

raoulbhatia commented:
FYI We have rebased the current patch on behalf of @peterhoneder onto latest master to ensure we are up2date

thecmdradama commented:
Hi guys, any progress on getting this PR merged in? @mloiseleur

raoulbhatia commented:
/remove-lifecycle stale

foyerunix commented:
Hello,

It seem I duplicated a great part of your effort in my own PR that also handle the reverse PTR creation: https://github.com/kubernetes-sigs/external-dns/pull/4782

Perhaps there is a way to merge the two into something that could be accepted ?

Best Regards.

peterhoneder commented:
> Hello,
> 
> It seem I duplicated a great part of your effort in my own PR that also handle the reverse PTR creation: #4782
> 
> Perhaps there is a way to merge the two into something that could be accepted ?
> 
> Best Regards.

I would be happy to see both of the PRs merged, especially since we use this PR already for a long time in production.

szuecs commented:
As long as it's only enabled for a certain new feature flag, I will not block this pr. Same for the other mentioned one.

In general I don't think it is a good idea but I will not block IMO broken architectures (don't try to explain why it is not broken 😉). Some are likely not in control of these assumptions and cannot easily change them.

Happy new year everyone, hope we can get this through the line for you.

mloiseleur commented:
@peterhoneder Do that work for you ? It seems to me that since #4782 already implements it behind a feature flag, we should focus on this one. Anything I missed ?

peterhoneder commented:
> @peterhoneder Do that work for you ? It seems to me that since #4782 already implements it behind a feature flag, we should focus on this one. Anything I missed ?

I'm not completely sure why this would need a feature flag, since you specifically have to add the annotation to enable the PodIP to be used anyway? Seems like a double opt-in then for me. Could you explain?

foyerunix commented:
Hello @peterhoneder,

I believe my PR also accommodates your use case. As I understand it, using `--ignore-non-host-network-pods` without my second new option `--pod-source-domain` should achieve the desired results.

Do you see any improvements we could make to better handle your use case?

Best regards,

raoulbhatia commented:
@foyerunix et al – we implemented our requirement based on #4782 with `--no-ignore-non-host-network-pods`.
We are still in in a verification/monitoring mode, but overall things look good at this point.

We will keep an eye on our deployment.
If everything is well, we can close this PR in favor of #4782 next week.

Thanks @foyerunix for pushing this :-)

mloiseleur commented:
@peterhoneder Now that https://github.com/kubernetes-sigs/external-dns/pull/4782 has been merged, I understand that we can close this one. Feel free to re-open it if I missed something.
/close
